### PR TITLE
SOX-18 SNFLK view load

### PR DESCRIPTION
**Pull request by zajca**
Jira: SOX-18
KBC: https://github.com/keboola/connection/pull/4443

Before asking for review make sure that:

## Checklist

- [x] New client method(s) has tests
- [x] Apiary file is updated

## Release

  - [x] I gave the PR a proper label:
    * major (BC break)
    * minor (new feature)
    * patch (backwards compatible fix)
    * no release (just test changes)

### CODE REVIEW

File: phpstan-baseline.neon [hunk 1]
```diff
@@ -3855,6 +3855,21 @@ parameters:
...
+		-
+			message: "#^Parameter \\#1 \\$connection of method Keboola\\\\Test\\\\Backend\\\\Workspaces\\\\Backend\\\\SnowflakeWorkspaceBackendDBAL\\:\\:getDbConnectionSnowflakeDBAL\\(\\) expects array, mixed given\\.$#"
+			count: 1
+			path: tests/Backend/Workspaces/Backend/SnowflakeWorkspaceBackendDBAL.php
+
+		-
+			message: "#^Property Keboola\\\\Test\\\\Backend\\\\Workspaces\\\\Backend\\\\SnowflakeWorkspaceBackendDBAL\\:\\:\\$schema \\(string\\) does not accept mixed\\.$#"
+			count: 1
+			path: tests/Backend/Workspaces/Backend/SnowflakeWorkspaceBackendDBAL.php
+
```
Comment (zajca) [inline phpstan-baseline.neon:3872]: tady to je na prd, ale když sem tam začal doplňovat to schéma tak bych to musel udělat všude jinak bych do baseline přidal asi dalších 1000 záznamů :(

File: tests/Backend/Workspaces/Backend/SnowflakeWorkspaceBackendDBAL.php [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+use Keboola\TableBackendUtils\Column\ColumnCollection;
+use Keboola\TableBackendUtils\Column\Snowflake\SnowflakeColumn;
+use Keboola\TableBackendUtils\Escaping\Snowflake\SnowflakeQuote;
+use Keboola\TableBackendUtils\Schema\Snowflake\SnowflakeSchemaReflection;
+use Keboola\TableBackendUtils\Table\Snowflake\SnowflakeTableQueryBuilder;
+use Keboola\TableBackendUtils\Table\Snowflake\SnowflakeTableReflection;
+use Keboola\TableBackendUtils\View\Snowflake\SnowflakeViewReflection;
+use Keboola\Test\Backend\WorkspaceConnectionTrait;
+
+class SnowflakeWorkspaceBackendDBAL implements WorkspaceBackend
```
Comment (zajca) [inline tests/Backend/Workspaces/Backend/SnowflakeWorkspaceBackendDBAL.php:16]: prakticky c&p původní SNFLK backend třídy. + nějaké reflection věci co sem potřeboval.

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 3]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+//        $fileId = $this->workspaceSapiClient->uploadFile(
+//            (new CsvFile($importFile))->getPathname(),
+//            (new FileUploadOptions())
+//                ->setNotify(false)
+//                ->setIsPublic(false)
+//                ->setCompress(true)
+//                ->setTags(['test-file-1'])
+//        );
+//        // load data from file workspace Not supported yet on S3 on ABS and not used in SNFLK
+//        $fileWorkspace = $workspaces->createWorkspace(
```
Comment (zajca) [inline tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:1029]: tady to sem nedělal musel bych to omezit na ABS a nechtěLo sa mě do toho drbat, asi to aj funguje, ale nechal bych to na potom.

Comment (tomasfejfar) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:1029]: Nebo `if (Něco::EDGE_CASES_TESTS_ENABLED)` abysme to pak nakonci mohli všechno zapnout a nemuseli hledat kde všude máme něco zacommentované?

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 4]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+        $tableRef = $backend->getTableReflection('languages');
+        self::assertEquals(['id', 'name', '_timestamp'], $tableRef->getColumnsNames());
+        try {
+            $backend->fetchAll('languages');
+            $this->fail('Must throw exception view columns mismatch');
+        } catch (Throwable $e) {
+            $this->assertStringContainsString('View columns mismatch with view definition for view \'languages\'', $e->getMessage());
+        }
+
+        // clear and create table again
```
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:905]: A to mazání je fajn, že se tím testuje, že WS má přístup k tomu viewčku.

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 5]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+        self::assertCount(0, $schemaRef->getTablesNames());
+        // view is still in workspace but not working
+        self::assertCount(1, $schemaRef->getViewsNames());
+        try {
+            $backend->fetchAll('languages');
+            $this->fail('View should not work after table drop');
+        } catch (Throwable $e) {
+            $this->assertStringContainsString('does not exist or not authorized', $e->getMessage());
+        }
+    }
```
Comment (tomasfejfar) [inline tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:1069]: Nevím, jestli bysme měli řešit nějaké edge cases - třeba co když si v transformaci to viewčko upravím? Měl by na to mít WS user právo? Řekl bych, že ne. Může ho smazat, může z něj selectovat, ale tím by to mělo hasnout. 

Ale když jsme zjistili, že stejně potřebuje ROIM, tak to moc nepomůže vynucení lineage - dycky to user může workaroundnout tím, že se podívá na tabulku přímo... hmmm, tak nic.

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (zajca) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:912]: ono to tam je pak že to přepíše, ale tady to bude pěkně vidět, tak sem to přesunul.

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 6]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+                    'source' => $tableId,
+                    'destination' => 'languages',
+                    'useView' => true,
+                ],
+            ],
+        ];
+
+        $workspaces->loadWorkspaceData($workspace['id'], $options);
+
+        $tableRef = $backend->getTableReflection('languages');
```
[the remark this answers is not in the document]
Comment (zajca) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:876]: přidal sem

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 7]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+                $currentToken['owner']['name'],
+                $currentToken['owner']['id']
+            ));
+        }
+
+        $bucketId = $this->getTestBucketId(self::STAGE_IN);
+
+        $workspaces = new Workspaces($this->workspaceSapiClient);
+        $workspace = $workspaces->createWorkspace([], true);
+        $backend = WorkspaceBackendFactory::createWorkspaceForSnowflakeDbal($workspace);
```
[the remark this answers is not in the document]
Comment (zajca) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:856]: tak ta normální sa menuje createWorkspaceBackend takže sem to udělal podle toho ať sa každá nemenuje jinak.

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php (hunk 3 above)
Comment (zajca) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php]: přidal sem

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php [hunk 8]
```diff
@@ -836,4 +836,235 @@ public function dataTypesDiffDefinitions()
...
+                    'useView' => false,
+                ],
+            ],
+            'preserve' => true,
+        ];
+        try {
+            $workspaces->loadWorkspaceData($workspace['id'], $options);
+            self::fail('Incremental load to view cannot work.');
+        } catch (ClientException $e) {
+            // this is expected edge case, view has also _timestamp col
```
[the remark this answers is not in the document]
Comment (zajca) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:976]: https://keboola.atlassian.net/browse/SOX-76

File: tests/Backend/Workspaces/WorkspacesSnowflakeTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomasfejfar) [reply tests/Backend/Workspaces/WorkspacesSnowflakeTest.php:859]: Tak pak `createWorkspacebackendForSnowflakeDbal` ne?

**Review by tomasfejfar — APPROVED** [review]
